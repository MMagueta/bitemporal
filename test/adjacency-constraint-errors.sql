\set QUIET 'on'
\pset footer 'off'

drop schema if exists ac cascade;
create schema ac;

\echo 'bitemporal.adjacency_constraint_errors - [ok]'
\pset title 'ac.ok'
create table ac.ok
    ( id                int not null
    , id2               int not null
    , value             int not null
    , value2            int not null
    , valid_period      int4range not null
    , system_period     int4range not null
    , primary key(id, id2, valid_period)
    , exclude using gist (id with =, id2 with =, valid_period with &&)
    , exclude using gist (id with =, id2 with =, value with =, value2 with =, valid_period with -|-) );

select * from bitemporal.adjacency_constraint_errors('ac.ok');

\echo 'bitemporal.adjacency_constraint_errors - [missing adjacency constraint]'
\pset title 'ac.missing'
create table ac.missing
    ( id                int not null
    , id2               int not null
    , value             int not null
    , value2            int not null
    , valid_period      int4range not null
    , system_period     int4range not null
    , primary key(id, id2, valid_period));

select * from bitemporal.adjacency_constraint_errors('ac.missing');

\echo 'bitemporal.adjacency_constraint_errors - [mismatched fields]'
\pset title 'ac.mismatched_fields'
create table ac.mismatched_fields
    ( id                int not null
    , id2               int not null
    , value             int not null
    , value2            int not null
    , valid_period      int4range not null
    , system_period     int4range not null
    , primary key(id, id2, valid_period)
    , exclude using gist (id with =, id2 with =, valid_period with &&)
    , exclude using gist (id with =, id2 with =, value with =, system_period with =, valid_period with -|-) );

select * from bitemporal.adjacency_constraint_errors('ac.mismatched_fields');

\echo 'bitemporal.adjacency_constraint_errors - [mismatched fields]'
\pset title 'ac.mismatched_fields_2'
create table ac.mismatched_fields_2
    ( id                int not null
    , id2               int not null
    , value             int not null
    , value2            int not null
    , valid_period      int4range not null
    , system_period     int4range not null
    , primary key(id, id2, valid_period)
    , exclude using gist (id with =, id2 with =, valid_period with &&)
    , exclude using gist (id with =, id2 with =, value with =, valid_period with -|-) );

select * from bitemporal.adjacency_constraint_errors('ac.mismatched_fields_2');

\echo 'bitemporal.adjacency_constraint_errors - [wrong operator on common field]'
\pset title 'ac.wrong_operator'
create table ac.wrong_operator
    ( id                int not null
    , id2               int not null
    , value             int not null
    , value2            int not null
    , valid_period      int4range not null
    , system_period     int4range not null
    , primary key(id, id2, valid_period)
    , exclude using gist (id with =, id2 with =, valid_period with &&)
    , exclude using gist (id with =, id2 with <>, value with =, value2 with =, valid_period with -|-) );

select * from bitemporal.adjacency_constraint_errors('ac.wrong_operator');

\echo 'bitemporal.adjacency_constraint_errors - [wrong operator on valid_time field]'
\pset title 'ac.wrong_operator_2'
create table ac.wrong_operator_2
    ( id                int not null
    , id2               int not null
    , value             int not null
    , value2            int not null
    , valid_period      int4range not null
    , system_period     int4range not null
    , primary key(id, id2, valid_period)
    , exclude using gist (id with =, id2 with =, valid_period with &&)
    , exclude using gist (id with =, id2 with =, value with =, value2 with =, valid_period with &&) );

select * from bitemporal.adjacency_constraint_errors('ac.wrong_operator_2');

\echo 'bitemporal.adjacency_constraint_errors - [multiple wrong operators]'
\pset title 'ac.wrong_operator_3'
create table ac.wrong_operator_3
    ( id                int not null
    , id2               int not null
    , value             int not null
    , value2            int not null
    , valid_period      int4range not null
    , system_period     int4range not null
    , primary key(id, id2, valid_period)
    , exclude using gist (id with =, id2 with =, valid_period with &&)
    , exclude using gist (id with <>, id2 with <>, value with <>, value2 with <>, valid_period with &&) );

select * from bitemporal.adjacency_constraint_errors('ac.wrong_operator_3');

\echo 'bitemporal.adjacency_constraint_errors - [multiple constraints, one ok]'
\pset title 'ac.multiple_one_ok'
create table ac.multiple_one_ok
    ( id                int not null
    , id2               int not null
    , value             int not null
    , value2            int not null
    , valid_period      int4range not null
    , system_period     int4range not null
    , primary key(id, id2, valid_period)
    , exclude using gist (id with =, id2 with =, valid_period with &&)
    , exclude using gist (id with <>, id2 with =, value with =, value2 with <>, valid_period with &&)
    , exclude using gist (id with =, valid_period with -|-, id2 with =, value2 with =, value with =) );

select * from bitemporal.adjacency_constraint_errors('ac.multiple_one_ok');

\echo 'bitemporal.adjacency_constraint_errors - [multiple constraints with primary key fields, none ok]'
\pset title 'ac.wrong_operator_none_ok'
create table ac.multiple_none_ok
    ( id                int not null
    , id2               int not null
    , value             int not null
    , value2            int not null
    , valid_period      int4range not null
    , system_period     int4range not null
    , primary key(id, id2, valid_period)
    , exclude using gist (id with =, id2 with =, valid_period with &&)
    , exclude using gist (id with <>, id2 with =, value with =, value2 with <>, valid_period with &&)
    , exclude using gist (id with =, valid_period with -|-, id2 with <>, value2 with =, value with =) );

select * from bitemporal.adjacency_constraint_errors('ac.multiple_none_ok');
