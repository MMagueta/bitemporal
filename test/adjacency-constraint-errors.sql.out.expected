                                     ac.ok
 namespace | relation | constraint_name | attribute | message | expected | was 
-----------+----------+-----------------+-----------+---------+----------+-----

                                                 ac.missing
 namespace | relation | constraint_name | attribute |                message                | expected | was 
-----------+----------+-----------------+-----------+---------------------------------------+----------+-----
 ac        | missing  |                 |           | Missing exclude adjacency constraint. |          | 

                                                 ac.mismatched_fields
 namespace |     relation      | constraint_name | attribute |                message                | expected | was 
-----------+-------------------+-----------------+-----------+---------------------------------------+----------+-----
 ac        | mismatched_fields |                 |           | Missing exclude adjacency constraint. |          | 

                                                 ac.mismatched_fields_2
 namespace |      relation       | constraint_name | attribute |                message                | expected | was 
-----------+---------------------+-----------------+-----------+---------------------------------------+----------+-----
 ac        | mismatched_fields_2 |                 |           | Missing exclude adjacency constraint. |          | 

                                                         ac.wrong_operator
 namespace |    relation    |                   constraint_name                    | attribute |      message      | expected | was 
-----------+----------------+------------------------------------------------------+-----------+-------------------+----------+-----
 ac        | wrong_operator | wrong_operator_id_id2_value_value2_valid_period_excl | id2       | Invalid operator. | =        | <>

                                                            ac.wrong_operator_2
 namespace |     relation     |                    constraint_name                     |  attribute   |      message      | expected | was 
-----------+------------------+--------------------------------------------------------+--------------+-------------------+----------+-----
 ac        | wrong_operator_2 | wrong_operator_2_id_id2_value_value2_valid_period_excl | valid_period | Invalid operator. | -|-      | &&

                                                            ac.wrong_operator_3
 namespace |     relation     |                    constraint_name                     |  attribute   |      message      | expected | was 
-----------+------------------+--------------------------------------------------------+--------------+-------------------+----------+-----
 ac        | wrong_operator_3 | wrong_operator_3_id_id2_value_value2_valid_period_excl | id           | Invalid operator. | =        | <>
 ac        | wrong_operator_3 | wrong_operator_3_id_id2_value_value2_valid_period_excl | id2          | Invalid operator. | =        | <>
 ac        | wrong_operator_3 | wrong_operator_3_id_id2_value_value2_valid_period_excl | value        | Invalid operator. | =        | <>
 ac        | wrong_operator_3 | wrong_operator_3_id_id2_value_value2_valid_period_excl | value2       | Invalid operator. | =        | <>
 ac        | wrong_operator_3 | wrong_operator_3_id_id2_value_value2_valid_period_excl | valid_period | Invalid operator. | -|-      | &&

                              ac.multiple_one_ok
 namespace | relation | constraint_name | attribute | message | expected | was 
-----------+----------+-----------------+-----------+---------+----------+-----

                                                         ac.wrong_operator_none_ok
 namespace |     relation     |                    constraint_name                     |  attribute   |      message      | expected | was 
-----------+------------------+--------------------------------------------------------+--------------+-------------------+----------+-----
 ac        | multiple_none_ok | multiple_none_ok_id_id2_value_value2_valid_period_excl | id           | Invalid operator. | =        | <>
 ac        | multiple_none_ok | multiple_none_ok_id_id2_value_value2_valid_period_excl | value2       | Invalid operator. | =        | <>
 ac        | multiple_none_ok | multiple_none_ok_id_id2_value_value2_valid_period_excl | valid_period | Invalid operator. | -|-      | &&
 ac        | multiple_none_ok | multiple_none_ok_id_valid_period_id2_value2_value_excl | id2          | Invalid operator. | =        | <>

